### DISCUSSION TITLE
Is MLIR compulsory for xDSL?

### DISCUSSION THREAD

**Discussion by ravil-mobile** [xdslproject/xdsl#1277] [General]
Hi All,

I just want to get some clarification on whether xDSL can operate without MLIR? According to your documentation, xDLS was inspired by MLIR (in terms of dialects, operations, attributes, etc.) and can work with it but not by default.

This aspect is important for me because having MLIR as a dependency is too cumbersome for a relatively small DSL which I would like to implement with xDSL.

I would really appreciate if somebody can answer. 

Thanks in advance!

**Comment by webmiche** [comment]
The short answer: Yes, xDSL can be used entirely without MLIR. There is no dependency in terms of code.

For the long answer I would need more details on your project. The main way to generate highly performant code with xDSL is through MLIR/LLVM at the moment, but this is not a fundamental problem, but rather a problem of engineering costs (i.e. no one has implement all the necessary things in xDSL yet).

**Reply by ravil-mobile** [reply]
Hi @webmiche. Many thanks for your quick reply.

My main idea is to generate C++/CUDA/HIP/SYCL source code for a given tensor expression. Then, the generated code and the main application can be compiled together. The code generator will be called directly from CMake. This approach has already been established in [SeisSol](https://github.com/SeisSol/SeisSol) project.

If I understood the documentation correctly, I can 1) import xDSL; 2) define my tensor-specific dialects with their operations and attributes; and, then, 3) use them. It seems to be ok for me in terms of the cost of maintenance. Alternatively, I could take advantage of some similar MLIR dialects but, as I said before, it is bulky for my specific case.

What do you think?

**Reply by webmiche** [reply]
I think this is exactly one of the kind of usecases we are targetting with xDSL. We work closely with some of the [devito](https://www.devitoproject.org/) and [psyclone](https://psyclone.readthedocs.io/en/stable/) people. I must say, I am not quite up to speed with how the current code generation works in the two from the xDSL side (probably @georgebisbas and @mesham know it better), but AFAIR, they are generating C/C++ through xDSL, so that seems quite similar to what you have in mind.

We will probably have to flesh out the exact details, and for the usecase you need it exactly. For papers, the dependency on MLIR as a proof of concept can be fine, but to distribute to end users, you don't want that. Devito has the exact same problem, so I expect there to be further development/support in that direction from the xDSL side.

Also, it might make sense to look over the xDSL and MLIR dialects and reuse parts.

**Reply by superlopuh** [reply]
We have a bunch of ways to run code generated by xDSL, but no way to compile to assembly for any specific hardware. Even if we lower to the LLVM dialect, we still need mlir-opt to generate LLVM IR and pass it on to LLVM from there. I don't know if this helps but at least with xDSL you don't need to distribute MLIR source, only a binary, which might make it ok to have as a dependency in your case?

**Reply by ravil-mobile** [reply]
Thanks @superlopuh,

I think the generation of CUDA/HIP/SYCL source code is the best for us. In this case, we can delegate low-level optimizations (e.g., instruction scheduling) to native compilers - i.e., `nvcc`, `amdclang`, `dpcpp` - which tend to do a better job than LLVM.

Our main project, SeisSol, has its DSL (called YATeTo), but it used to be quite CPU-centric. Later, GPU support was added. Looking at the source code now, we can see that it is better to change the software design. We are planning the source code refactoring and looking around to see whether there are any other projects that we can employ.

I remember there was a talk at the SIAM-CSE@2023 conference dedicated to xDSL. Unfortunately, I didn't manage to talk to the presenter.
